### PR TITLE
Emit multiple fields from a runtime field script

### LINKED ISSUE #68203: Add support for emitting multiple fields values from a script
We recently added support for runtime fields, that are computed at search time based on a painless script. As of today, a runtime field script can emit values for a single field, the one that the script is declared under.

We would like to add the ability for a script to emit values for multiple fields. This will be achieved by introducing support for a new field type (name to be defined) as part of the runtime section. Its script emits fields that belong to such object. This is  particularly useful given that scripts support grok and dissect (#68088):

```
PUT localhost:9200/logs/_mappings
{
  "runtime" : {
    "log" : {
      "type" : "tbd",
      "script": '''
        emit(grok('%{COMMONAPACHELOG}').extract(doc["message"].value)));
      ''',
      "fields" : {
        "clientip" : {
          "type" : "ip"
        },
        "verb" : {
          "type" : "keyword"
        },
        "request" : {
          "type" : "keyword"
        },
        "response" : {
          "type" : "long"
        }
      }
    }
  },
  "properties" : {
    "message" : {
      "type" : "keyword"
    }
  }
}
```

In the example above, the `grok` function splits the message field into sub-fields based on the provided grok pattern, and each of the resulting fields is emitted in the following loop. The emitted fields need to be listed under the sub-fields in order to specify their type and make them searchable (and discoverable through field_caps) like any other field:

```
POST /logs*/_search
{
  "aggs": {
    "response_codes": {
      "range": {
        "field": "log.response",
        "ranges": [
          { "to": 300 },
          { "from": 300, "to": 400 },
          { "from": 500 }
        ]
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by javanna**
We have recently introduced support for grok and dissect to the runtime fields Painless context that allows to split a field into multiple fields. Though each runtime field can only emit values for a single field. This PR introduces support for emitting multiple fields from the same script.

The API call to define a runtime field that emits multiple fields is the following:

```
PUT localhost:9200/logs/_mappings
{
    "runtime" : {
      "log" : {
        "type" : "composite",
        "script" : "emit(grok(\"%{COMMONAPACHELOG}\").extract(doc[\"message.keyword\"].value))",
        "fields" : {
            "clientip" : {
                "type" : "ip"
            },
            "response" : {
                "type" : "long"
            }
        }
      }
    }
}
```

The script context for this new field type accepts two emit signatures:

```
emit(String, Object)
emit(Map)
```

Sub-fields need to be declared under `fields` in order to be discoverable through the field_caps API and accessible through the search API. In this first iteration, mapping additional sub-fields requires to provide the whole object with its script and all the existing plus new sub-fields. In a follow-up we will address this by allowing users to provide only what needs to be changed or added.

The way that it emits multiple fields is by returning multiple `MappedFieldType`s from `RuntimeField#asMappedFieldTypes`. The sub-fields are instances of the runtime fields that are already supported, with a little tweak to adapt the script defined by their parent to an artificial script factory for each of the sub-fields that makes its corresponding sub-field accessible. This approach allows to reuse all of the existing runtime fields code for the sub-fields.

The runtime section has been flat so far as it has not supported objects until now. That stays the same, meaning that runtime fields can have dots in their names. Because there are though two ways to create the same field with the introduction of the ability to emit multiple fields, we have to make sure that a runtime field with a certain name cannot be defined twice, which is why the following mappings are rejected with the error `Found two runtime fields with same name [log.response]`:

```
PUT localhost:9200/logs/_mappings
{
    "runtime" : {
        "log.response" : {
            "type" : "keyword"
        },
        "log" : {
            "type" : "composite",
            "script" : "emit(\"response\", grok(\"%{COMMONAPACHELOG}\").extract(doc[\"message.keyword\"].value)?.response)",
            "fields" : {
                "response" : {
                    "type" : "long"
                }
            }
        }
    }
}
```

Closes #68203

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    public static final Parser PARSER = new Parser(name ->
+        new RuntimeField.Builder(name) {
+            private final FieldMapper.Parameter<Script> script = new FieldMapper.Parameter<>(
+                "script",
+                false,
+                () -> null,
+                AbstractScriptFieldType.Builder::parseScript,
+                RuntimeField.initializerNotSupported()
+            ).setValidator(s -> {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java:39]: it seems like this validator does nothing, I need to look into it. @romseygeek any idea?

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        builder.startObject(name);
+        builder.field("type", "object");
+        boolean includeDefaults = params.paramAsBoolean("include_defaults", false);
+        for (FieldMapper.Parameter<?> parameter : parameters) {
+            parameter.toXContent(builder, includeDefaults);
+        }
+        builder.startObject("fields");
+        for (RuntimeField subfield : subfields) {
+            subfield.toXContent(builder, params);
+        }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java:108]: we print out all the parameters above, and fields is also a parameter, but nothing gets printed out. It is probably good as we can only parse RuntimeFields at a later time, which is what we need to print out the subfields. Is what I am doing here ok? Do we have to somehow make sure that serializing the parameters does not serialize also the fields, why does nothing happen at the moment @romseygeek ?

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java:39]: It will throw an error if there's no script defined on the object field, no?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java:39]: that's what I'd expect it to do, but I meant to bring up that it has no effect.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectRuntimeField.java:39]: I tracked it. we no longer call validate on the parameters. It is no problem as we never use a validator in the existing parameters. I will fix this

**Comment by romseygeek** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * Runs the object script and returns the values that were emitted for the provided field name
+     * @param field the field name to extract values from
+     * @return the values that were emitted for the provided field
+     */
+    public final List<Object> getValues(String field) {
+        //TODO for now we re-run the script every time a leaf field is accessed, but we could cache the values?
+        fieldValues.clear();
+        execute();
+        return fieldValues.get(field);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java:49]: should fieldValues be cleared before returning field so that it does not hold onto unnecessary state?

**Review by ywelsch — COMMENTED** [review]
I've reviewed this as someone not familiar with this part of the codebase (so needs another reviewer). I've left a few questions / comments but overall the PR looks in good shape to me.

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        this.subfields = subfields;
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public Collection<MappedFieldType> asMappedFieldTypes() {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java:100]: Maybe return a `Stream` from this if you are going to be `flatMap`ing it. If called needs a list they can make one?

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+ */
+public class CompositeRuntimeField implements RuntimeField {
+
+    public static final String CONTENT_TYPE = "composite";
+
+    public static final Parser PARSER = new Parser(name ->
+        new RuntimeField.Builder(name) {
+            private final FieldMapper.Parameter<Script> script = new FieldMapper.Parameter<>(
+                "script",
+                false,
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java:37]: Runtime field definitions don't get merged, they get replaced entirely, so this parameter doesn't apply.

File: server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        super(fieldName, params, searchLookup, ctx);
+    }
+
+    /**
+     * Runs the object script and returns the values that were emitted for the provided field name
+     * @param field the field name to extract values from
+     * @return the values that were emitted for the provided field
+     */
+    public final List<Object> getValues(String field) {
+        //TODO for now we re-run the script every time a leaf field is accessed, but we could cache the values?
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java:46]: Definitely.  There aren't any docs in this PR, they will be added as a follow-up in collaboration with the docs team, but we'll make sure to call out this limitation.

**Review by nik9000 — COMMENTED** [review]
Oh boy!

I wonder if we could have `emit` on the composite script bind its results into all of the "child" scripts immediately. Right now it looks like you can run a script and it'll collect a `Map` of values. And then, after the script is finished, you extract just the ones you need. That's good because you don't convert the ones you don't need. But its bad that you silently ignore when you emit invalid values that you aren't reading. It also feels bad that the error comes after the script instead of *during* it. If we throw the error while running the script we'd get the line it fails on. Painless has `try/catch` so you could even use that to dodge error cases or add extra information when an error is thrown.

A smaller thing, but now that there are three sorts of subclasses for stuff like `DoubleFieldScript` maybe it'd be a good idea to have a more natural shaped supertype. The design of these classes is very much around painless implementing them, but we have those two cases where we implement it by hand. It might make sense to make refactor the base classes so they are less "painless shaped" and make a "painless shaped" layer that it can implement. I'm not sure, and, regardless, I don't think that is a "for now" thing.

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java:37]: Reusing the field mapper parameter parsing was a nice hack but this shows one of the places where its kind of confusing. I'm not sure what to do about it though.

**Comment by romseygeek** [comment]
> I wonder if we could have emit on the composite script bind its results into all of the "child" scripts immediately. Right now it looks like you can run a script and it'll collect a Map of values. And then, after the script is finished, you extract just the ones you need. That's good because you don't convert the ones you don't need. But its bad that you silently ignore when you emit invalid values that you aren't reading. It also feels bad that the error comes after the script instead of during it. If we throw the error while running the script we'd get the line it fails on. Painless has try/catch so you could even use that to dodge error cases or add extra information when an error is thrown.

At the moment the script is actually run by the child field, and there's no caching or sharing of values between different child fields.  So if you have a grok defined on a parent field 'message' and you search on 'message.status' and 'message.ip', both fields will run the painless script separately and then extract the part of the map that they need.  This isn't ideal, as Yannick mentions, but it's the way runtime fields work now.  It would be great to rework things so that different runtime fields could share a LeafSearchLookup, in which case we could cache these results there, but that's a much bigger project.

File: server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java [hunk 7]
```diff
@@ -92,25 +99,33 @@ public Parser(Function<String, RuntimeField.Builder> builderFunction) {
...
      * @param parserContext the parser context that holds info needed when parsing mappings
+     * @param parent the name of the parent field that the field being parsed is part of, if any
+     * @param parentScriptFactory script factory of the parent field that the field being parsed is part of, if any
      * @param supportsRemoval whether a null value for a runtime field should be properly parsed and
      *                        translated to the removal of such runtime field
      * @return the parsed runtime fields
      */
     static Map<String, RuntimeField> parseRuntimeFields(Map<String, Object> node,
                                                         MappingParserContext parserContext,
+                                                        String parent,
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java:127]: Since 'parent' and 'parentScriptFactory' only apply to `composite` field parsing, it'd be nice to avoid exposing them in these very general methods. If we created a dedicated parsing method for composite fields, could we avoid updating `RuntimeField#parseRuntimeFields` and `RuntimeField#parse`?

**Review by jtibshirani — COMMENTED** [review]
This makes sense to me overall, I left some initial comments. The name `composite` has really grown on me!

File: server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java [hunk 8]
```diff
@@ -78,6 +82,9 @@ public final void parse(String name, MappingParserContext parserContext, Map<Str
                 parameter.parse(name, parserContext, propNode);
                 iterator.remove();
             }
+            for (Parameter<?> parameter : getParameters()) {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java:85]: Yes, that's correct. We also didn't actually need to validate any of the parameters before either though, so it was an oversight rather than a bug.

File: server/src/main/java/org/elasticsearch/index/mapper/LeafRuntimeField.java [hunk 9]
```diff
@@ -16,8 +16,9 @@
 import java.util.List;
 
 /**
- * RuntimeField base class for leaf fields that will only ever return
- * a single MappedFieldType from {@link RuntimeField#asMappedFieldTypes()}
+ * RuntimeField base class for leaf fields that will only ever return a single {@link MappedFieldType}
+ * from {@link RuntimeField#asMappedFieldTypes()}. Can be a standalone runtime field, or part of an object runtime field,
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/LeafRuntimeField.java:20]: I think this is a leftover, I'll change this so that it always returns the full name including the parent prefix.

**Comment by romseygeek** [comment]
Thanks for the reviews everybody!

I have reworked things so that runtime builders now have two methods, one to create a simple leaf field and one to create a field with a parent.  The parsing code can take an optional function that will choose which method to call when building fields, and it nicely separates out composite field construction from 'normal' field construction.

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractScriptFieldType.java [hunk 10]
```diff
@@ -194,33 +195,56 @@ protected final LeafFactory leafFactory(SearchExecutionContext context) {
...
+            }
+            String fullName = parent + "." + name;
+            return new LeafRuntimeField(
+                name,
+                createFieldType(fullName, getCompositeLeafFactory(parentScriptFactory), getScript(), meta()),
+                getParameters()
+            );
+        }
+
+        final RuntimeField createRuntimeField(String parent, Factory scriptFactory) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/AbstractScriptFieldType.java]: I think 'parent' is always `null` here, so we could remove it (keeping any mention of 'parents' confined to `createChildRuntimeField`).

File: server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+//TODO expose this context to the painless execute API
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java:19]: Can this TODO be removed?

File: server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java [hunk 12]
```diff
@@ -635,8 +635,12 @@ public Index getFullyQualifiedIndex() {
         if (runtimeMappings.isEmpty()) {
             return Collections.emptyMap();
         }
-        Map<String, RuntimeField> runtimeFields = RuntimeField.parseRuntimeFields(new HashMap<>(runtimeMappings),
-            mapperService.parserContext(), false);
+        //TODO add specific tests to SearchExecutionTests similar to the ones in FieldTypeLookupTests
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:638]: Checking I'm not missing something, this is not really related to the change?

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public Stream<MappedFieldType> asMappedFieldTypes() {
+        return subfields.stream().flatMap(RuntimeField::asMappedFieldTypes);
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java:105]: Thinking out loud, would it make sense to compute the mapped field types up-front as we do for `LeafRuntimeField`? This might less us avoid changing the API to `Stream` (which seems much less common than using `Collection`).

**Review by jtibshirani — COMMENTED** [review]
This is looking good to me, just left some small comments. As a caveat, I am not deeply familiar with this code and its history (`AbstractScriptFieldType`, `RuntimeField`, etc.)

File: server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java (hunk 12 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:638]: No, I think this was just a reminder that Luca added because test coverage here isn't great

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractScriptFieldType.java (hunk 10 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/AbstractScriptFieldType.java]: ++, removed in 767633d

File: server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java (hunk 11 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/script/CompositeFieldScript.java:19]: It was still a TODO but it turned out to be very straightforward to add support for it, so I have implemented it (and removed the TODO!)

File: server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java (hunk 13 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/CompositeRuntimeField.java:105]: I gave this a try but I don't think it makes a difference to the API - `Stream` is still a useful thing to return here because of the validation that occurs in `RuntimeField.collectFieldTypes()`, and `asMappedFieldTypes()` is only called once per mapping update so we don't save anything particularly by building them up-front.  I think on balance I'd prefer to keep things working this way.

**Review by jtibshirani — APPROVED** [review]

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Comment by markharwood** [comment]
Hi @romseygeek I have a null pointer [fix for this in master](https://github.com/elastic/elasticsearch/issues/76716) but when I tried backporting that fix it [looks like](https://github.com/elastic/elasticsearch/search?q=%22Emit+multiple+fields+from+a+runtime+field+script%22&type=commits) this feature is not backported in 7.x?

**Comment by markharwood** [comment]
Ignore my last comment. I can see the NPE is in a different class in 7.15 - TransportFieldCapabilitiesIndexAction.

**Comment by mattkime** [comment]
@javanna - I'd like to confirm - composite fields can only have 'child' fields one level deep, correct?

**Comment by romseygeek** [comment]
@mattkime yes that's correct
